core/src/smartcontracts/wasm/mod.rs

```diff
@@ -643,149 +652,46 @@ impl<'wrld> Runtime<'wrld> {
-        Ok(obj)
-    }
-
-    /// Encode `bytes` to the given `memory` with the given `alloc_fn` and `context`
-    ///
-    /// Return the offset of the encoded object
-    fn encode_bytes_into_memory(
-        bytes: &[u8],
-        memory: &wasmtime::Memory,
-        alloc_fn: &wasmtime::TypedFunc<WasmUsize, WasmUsize>,
-        mut context: impl wasmtime::AsContextMut,
-    ) -> Result<WasmUsize, Trap> {
-        let mut encode = || -> eyre::Result<WasmUsize> {
-            bytes
-                .len()
-                .try_into()
-                .wrap_err("Bytes length is too big and can't be represented as `WasmUsize`")
-                .and_then(|len| alloc_fn.call(&mut context, len).map_err(Into::into))
-                .and_then(|offset| {
-                    let offset_usize = offset
-                        .try_into()
-                        .wrap_err("Offset is too big and can't be represented as `usize`")?;
-                    memory.write(&mut context, offset_usize, bytes)?;
-
-                    Ok(offset)
-                })
-        };
-        encode().map_err(|error| Trap::new(error.to_string()))
-    }
-
-    /// Encode the given object but also add it's length in front of it. This can be considered
-    /// a custom encoding format
-    ///
-    /// Usually, to retrieve the encoded object both pointer and the length of the allocation
-    /// are provided. However, due to the lack of support for multivalue return values in stable
-    /// `WebAssembly` it's not possible to return two values from a wasm function without some
-    /// shenanignas. In those cases, only one value is sent which is pointer to the allocation
-    /// with the first element being the length of the encoded object following it.
-    fn encode_with_length_prefix<T: Encode>(obj: &T) -> Result<Vec<u8>, Trap> {
-        let len_size_bytes = core::mem::size_of::<WasmUsize>();
-
-        let mut r = Vec::with_capacity(len_size_bytes + obj.size_hint());
-
-        // Reserve space for length
-        r.resize(len_size_bytes, 0);
-        obj.encode_to(&mut r);
-
-        // Store length as byte array in front of encoding
-        let len = &WasmUsize::try_from(r.len()).map_err(|e| Trap::new(e.to_string()))?;
-        r[..len_size_bytes].copy_from_slice(&len.to_le_bytes());
+        mut store: &mut wasmtime::Store<State<'wrld>>,
+    ) -> Result<wasmtime::Instance, InstantiationError> {
+        let instance = self
+            .linker
+            .instantiate(&mut store, module)
+            .map_err(|err| InstantiationError::Linker(eyre!(Box::new(err))))?;
+
+        // Check mandatory exports
+        let _ = Self::get_memory(&mut (&instance, &mut store))?;
+        let _ =
+            Self::get_typed_func::<WasmUsize, WasmUsize>(&instance, store, export::WASM_ALLOC_FN)?;
+        let _ = Self::get_typed_func::<(WasmUsize, WasmUsize), ()>(
+            &instance,
+            store,
+            export::WASM_DEALLOC_FN,
+        )?;
```

Comment:
isn't this done automatically by `wasmtime` when linking? I'm quite sure it will not allow you to instantiate a module with undefined imports. so it's redundant

Reply:
I meant to say this check is useless because this will never fail

Reply:
> I'm quite sure it will not allow you to instantiate a module with undefined imports. so it's redundant

How does wasmtime know which imports we require? Why isn't it possible to load a smartcontract without allocation function?

Reply:
got it